[MoE] Add deduplicated combine

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168047
* #167792
* __->__ #167676
* #167144
* #166896
* #166899
* #166898
* #166774
* #163815
* #163814
* #166772
* #165482
* #165101
* #164181
* #164164

Dedup combine kernel

Test intra-node combine

inter-node combine

Fix b_start

Works

Remove intermediate test

Move into separate file

MemPool works for combine

Add scatter_reduce_

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -1405,6 +1406,292 @@ void _all_to_all_v_2d_index_push(
+
+    // Accumulate the received tokens
+    constexpr int UNROLL = 4;
+    T acc[UNROLL] = {0};
```
Why not just std::array at this point?

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update on "[MoE] Add deduplicated combine"


Dedup combine kernel

Test intra-node combine

inter-node combine

Fix b_start

Works

Remove intermediate test

Move into separate file

MemPool works for combine

Add scatter_reduce_

[ghstack-poisoned]
